### PR TITLE
update libs

### PR DESCRIPTION
**Pull request by pfi79**
https://github.com/hyperledger/fabric/issues/3650

Libraries have been updated:

- github.com/hyperledger/fabric-chaincode-go/v2 v2.0.0
- github.com/hyperledger/fabric-protos-go-apiv2 v0.3.3
- github.com/hyperledger/fabric-config v0.3.0
- github.com/IBM/idemix v0.0.2-0.20240913182345-72941a5f41cd

From the features:

- proto.Marshal(nil) now does not return an error
- from proto to string, protobuf library inserts floating spaces into text and it is dangerous to compare with constant strings.
- it is almost impossible to break a change into small parts

There are floating bugs, especially in unit tests, which appear periodically

### CODE REVIEW

File: common/ledger/blkstorage/block_serialization.go
Comment (C0rWin) [inline common/ledger/blkstorage/block_serialization.go]: if function not going to return and error, I guess you can get rid of it in function signature also.

File: common/ledger/blkstorage/block_stream.go
Comment (C0rWin) [inline common/ledger/blkstorage/block_stream.go:109]: what is the use case for negative value of n?

File: common/ledger/blkstorage/blockfile_mgr.go
Comment (C0rWin) [inline common/ledger/blkstorage/blockfile_mgr.go]: itit will always return the nil error value, IMO you'd better remove it.

**Review by C0rWin — COMMENTED** [review]
@pfi79 thanks, this is a long standing technical debt we had it the project. I am fine with the change in general, not sure it's possible to scope it to iduce less changed files. 

need to carefully decided the order of how we would like to make a merge, cause it affects other repos.

regardless, I think additional couple of eyes needed to conclude on the review due to it's size (@denyeart @yacovm)

File: common/ledger/blkstorage/block_stream.go
Comment (pfi79) [reply common/ledger/blkstorage/block_stream.go:109]: The DecodeVarint method checks for a negative value of n. I just repeated it.

https://github.com/golang/protobuf/blob/master/proto/buffer.go#L41

File: common/ledger/blkstorage/block_serialization.go
Comment (pfi79) [reply common/ledger/blkstorage/block_serialization.go]: fix

File: common/ledger/blkstorage/blockfile_mgr.go
Comment (pfi79) [reply common/ledger/blkstorage/blockfile_mgr.go]: fix

File: common/ledger/blkstorage/block_stream.go
Comment (pfi79) [reply common/ledger/blkstorage/block_stream.go:109]: ConsumeVarint can return -1

**Comment by yacovm** [comment]
> @pfi79 thanks, this is a long standing technical debt we had it the project. I am fine with the change in general, not sure it's possible to scope it to iduce less changed files.
> 
> need to carefully decided the order of how we would like to make a merge, cause it affects other repos.
> 
> regardless, I think additional couple of eyes needed to conclude on the review due to it's size (@denyeart @yacovm)

Given Fabric v3.0 GA should be out very shortly, perhaps it would make sense to merge this afterwards? 
I don't think now is a good time be using temporary github forks for idemix and fabric-config given that we aim to cut the GA version very soon. 

@denyeart when are you planning to do the GA release?

**Comment by pfi79** [comment]
> > @pfi79 thanks, this is a long standing technical debt we had it the project. I am fine with the change in general, not sure it's possible to scope it to iduce less changed files.
> > need to carefully decided the order of how we would like to make a merge, cause it affects other repos.
> > regardless, I think additional couple of eyes needed to conclude on the review due to it's size (@denyeart @yacovm)
> 
> Given Fabric v3.0 GA should be out very shortly, perhaps it would make sense to merge this afterwards? I don't think now is a good time be using temporary github forks for idemix and fabric-config given that we aim to cut the GA version very soon.
> 
> @denyeart when are you planning to do the GA release?

Sorry, but changes to [fabric-config](https://github.com/hyperledger/fabric-config/pull/66) and [idemix](https://github.com/IBM/idemix/pull/47) have been proposed a week ago. 
If they are accepted, I will replace these versions here immediately.

And I was hoping they would be accepted sooner than this change.

**Comment by yacovm** [comment]
> > > @pfi79 thanks, this is a long standing technical debt we had it the project. I am fine with the change in general, not sure it's possible to scope it to iduce less changed files.
> > > need to carefully decided the order of how we would like to make a merge, cause it affects other repos.
> > > regardless, I think additional couple of eyes needed to conclude on the review due to it's size (@denyeart @yacovm)
> > 
> > 
> > Given Fabric v3.0 GA should be out very shortly, perhaps it would make sense to merge this afterwards? I don't think now is a good time be using temporary github forks for idemix and fabric-config given that we aim to cut the GA version very soon.
> > @denyeart when are you planning to do the GA release?
> 
> Sorry, but changes to [fabric-config](https://github.com/hyperledger/fabric-config/pull/66) and [idemix](https://github.com/IBM/idemix/pull/47) have been proposed a week ago. If they are accepted, I will replace these versions here immediately.
> 
> And I was hoping they would be accepted sooner than this change.

It's great that you have those ready, your maintenance work is appreciated.

Given you have already issued PRs for the module dependencies, it would make sense to merge them first before merging this.

**Comment by pfi79** [comment]
> > > > @pfi79 thanks, this is a long standing technical debt we had it the project. I am fine with the change in general, not sure it's possible to scope it to iduce less changed files.
> > > > need to carefully decided the order of how we would like to make a merge, cause it affects other repos.
> > > > regardless, I think additional couple of eyes needed to conclude on the review due to it's size (@denyeart @yacovm)
> > > 
> > > 
> > > Given Fabric v3.0 GA should be out very shortly, perhaps it would make sense to merge this afterwards? I don't think now is a good time be using temporary github forks for idemix and fabric-config given that we aim to cut the GA version very soon.
> > > @denyeart when are you planning to do the GA release?
> > 
> > 
> > Sorry, but changes to [fabric-config](https://github.com/hyperledger/fabric-config/pull/66) and [idemix](https://github.com/IBM/idemix/pull/47) have been proposed a week ago. If they are accepted, I will replace these versions here immediately.
> > And I was hoping they would be accepted sooner than this change.
> 
> It's great that you have those ready, your maintenance work is appreciated.
> 
> Given you have already issued PRs for the module dependencies, it would make sense to merge them first before merging this.

Can you review and comment on them?

**Comment by denyeart** [comment]
I agree dependent projects should be completed first. @ale-linux is the maintainer for idemix, @ale-linux can you look at these:
https://github.com/IBM/idemix/pull/47
https://github.com/IBM/idemix/pull/48

@pfi79 Thank you for adding some of the plan to https://github.com/hyperledger/fabric/issues/3650. Could you clarify in that issue the entire plan so that it is crystal clear? What is the complete set of PRs that will be needed across all repositories? Are they all open at this time? What tests will be required to mitigate risks associated with the update?

Based on the idemix progress and the overall plan, we can then make a decision about the release.
Ideally a major change like this would go in at a major release boundary. That would imply getting it in prior to cutting v3.0. However, we'll need to assess the overall scope and risk and progress in dependency projects.

File: go.mod
Comment (denyeart) [inline go.mod]: A proper v2.0.0 release is now available - https://github.com/hyperledger/fabric-chaincode-go/releases/tag/v2.0.0

**Comment by denyeart** [comment]
I've built Fabric components from this PR and tested them in a mixed environment with Fabric v2.5.9 nodes. Everything seems to work in a mixed environment.

Next step is for @ale-linux to review https://github.com/IBM/idemix/pull/47.

**Comment by denyeart** [comment]
@pfi79 I think you've made updates to the "delete shadow copy proto-struct" PR which diverges from the commit in this PR. I'm agreeable to having all the commits in this PR and closing the other PRs, if you can update everything here. That will make it cleaner for a single PR to test and merge. What do you think?

**Comment by pfi79** [comment]
@denyeart @C0rWin @yacovm 

All forks cleared, pips passed.
I suggest that you take another close look at the change. Because it doesn't depend on anything else.

**Comment by denyeart** [comment]
@pfi79 Thanks for the persistence and getting this done. I've tested latest commit again in a mixed environment with new orderer and peer nodes alongside v2.5.9 orderers and peer nodes. All looks good.

The changes look good to me. I'd like to get this in and then release v3.0 shortly. I'll give the others one more chance for final review before merging.

**Comment by C0rWin** [comment]
> @pfi79 Thanks for the persistence and getting this done. I've tested latest commit again in a mixed environment with new orderer and peer nodes alongside v2.5.9 orderers and peer nodes. All looks good.
> 
> The changes look good to me. I'd like to get this in and then release v3.0 shortly. I'll give the others one more chance for final review before merging.

@denyeart looks for me, IMO we can get it in.

**Comment by denyeart** [comment]
@pfi79 I see that you've updated SmartBFT in https://github.com/hyperledger-labs/SmartBFT/pull/597. Do you plan to update Fabric to use the updated SmartBFT? Does it matter either way? We are planning to release Fabric v3.0 shortly...

**Comment by pfi79** [comment]
> @pfi79 I see that you've updated SmartBFT in https://github.com/hyperledger-labs/SmartBFT/pull/597. Do you plan to update Fabric to use the updated SmartBFT? Does it matter either way? We are planning to release Fabric v3.0 shortly...

Yes, and smartbft and gomega, figured out the error. I thought to send them right after this pr. But if you think otherwise, then I can add them to this pr. although this pr is already very big.

**Comment by denyeart** [comment]
> > @pfi79 I see that you've updated SmartBFT in [hyperledger-labs/SmartBFT#597](https://github.com/hyperledger-labs/SmartBFT/pull/597). Do you plan to update Fabric to use the updated SmartBFT? Does it matter either way? We are planning to release Fabric v3.0 shortly...
> 
> Yes, and smartbft and gomega, figured out the error. I thought to send them right after this pr. But if you think otherwise, then I can add them to this pr. although this pr is already very big.

Ok thanks, I'll merge this PR and then you can do the follow-on PRs.

**Review by denyeart — APPROVED** [review]

**Pull request merged**
